include/zephyr/net/igmp.h

```diff
@@ -49,7 +49,36 @@ static inline int net_ipv4_igmp_join(struct net_if *iface,
 #endif
 
 /**
- * @brief Leave a given multicast group.
+ * @brief Join a given multicast group using IGMPv3.
+ *
+ * @param iface Network interface where join message is sent
+ * @param addr Multicast group to join
+ * @param include True to include source list, false to exclude
+ * @param sources_list List of sources to include or exclude
+ * @param source_len Length of source list
+ *
+ * @return Return 0 if joining was done, <0 otherwise.
+ */
+#if defined(CONFIG_NET_IPV4_IGMPV3)
+int net_ipv4_igmpv3_join(struct net_if *iface, const struct in_addr *addr, bool include,
```

Comment:
I created a separate API because IGMPv3 has some additional arguments/functionality. How would you add the filter groups to a request if the API remains unchanged? For leaving a group, Kconfig should do the job indeed.

Reply:
I'm not saying the API has to remain unchanged, I just put under discussion whether it makes sense to have a separate API for IGMPv3 only. Right now it's a bit counter-intuitive for me, that even after enabling IGMPv3 most (if not all) of the networking components would still send IGMPv2 reports. 

Just as an idea, `net_ipv4_igmp_join()` could accept an additional (optional) `struct igmp_param *` argument, with content dependent on the enabled IGMP version, with some reasonable fallback behavior if not provided (no filters for IGMPv3 if that makes sense). Net components that do not care about this additional functionality (is it just filtering or more?) could just skip the parameter.

Reply:
IGMPv3 does only support filters. I try your suggestion and see how that turns out.

Reply:
I agree with @rlubos, the public APIs should not need to mention v2/v3 etc. but be generic as the caller does not need to care what IGMP version we are using.

Reply:
I made the join/leave more generic. Now, the IGMP version is selected by the config. I added an optional param struct like @rlubos suggested.

Reply:
I will add extra tests and needed documentation for the change of the public API once the changes are on point.